File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/elm/ElmEditor.java
package org.cqframework.cql.cql2elm.elm;

import java.util.List;
import java.util.Objects;
import org.cqframework.cql.elm.tracking.Trackable;
import org.cqframework.cql.elm.utility.Visitors;
import org.cqframework.cql.elm.visiting.FunctionalElmVisitor;
import org.hl7.elm.r1.Element;
import org.hl7.elm.r1.Library;

public class ElmEditor {

    private final List<ElmEdit> edits;
    private final FunctionalElmVisitor<Trackable, Void> visitor;

    public ElmEditor(List<ElmEdit> edits) {
        this.edits = Objects.requireNonNull(edits);
        this.visitor = Visitors.from((elm, context) -> elm, this::applyEdits);
    }

    public void edit(Library library) {
        this.visitor.visitLibrary(library, null);
    }

    protected Trackable applyEdits(Trackable aggregate, Trackable nextResult) {
        if (nextResult instanceof Element) {
            for (ElmEdit edit : edits) {
                edit.edit((Element) nextResult);
            }
        }

        return aggregate;
    }
}

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/elm/ElmEditor.kt
package org.cqframework.cql.cql2elm.elm

import org.cqframework.cql.elm.tracking.Trackable
import org.cqframework.cql.elm.utility.Visitors
import org.cqframework.cql.elm.visiting.FunctionalElmVisitor
import org.hl7.elm.r1.Element
import org.hl7.elm.r1.Library

class ElmEditor(private val edits: List<IElmEdit>) {
    private val visitor: FunctionalElmVisitor<Trackable?, Void> =
        Visitors.from({ t, _ -> t }, { current, next -> this.aggregateResults(current, next) })

    fun edit(library: Library) {
        visitor.visitLibrary(library, null)

        // This is needed because aggregateResults is not called on the library itself.
        this.applyEdits(library)
    }

    private fun aggregateResults(aggregate: Trackable?, nextResult: Trackable?): Trackable? {
        applyEdits(nextResult)
        return aggregate
    }

    fun applyEdits(trackable: Trackable?) {
        if (trackable is Element) {
            for (edit in edits) {
                edit.edit(trackable)
            }
        }
    }
}
